File: source/c/async_file_processing.c
// gcc -o async_file_processing async_file_processing.c -lpthread
#include <stdio.h>
#include <pthread.h>
#include <string.h>

typedef struct {
    const char *name;
    int *data;
    int len;
    int sum;
    int max;
    int count;
} FileResult;

void *process_file(void *arg) {
    FileResult *r = (FileResult*)arg;
    r->sum = 0; r->max = r->data[0];
    for (int i = 0; i < r->len; i++) {
        r->sum += r->data[i];
        if (r->data[i] > r->max) r->max = r->data[i];
    }
    r->count = r->len;
    return NULL;
}

int main(void) {
    int data_a[] = {10, 20, 30, 40, 50};
    int data_b[] = {5, 15, 25, 35, 45, 55};
    int data_c[] = {100, 200, 300};

    FileResult results[3] = {
        {"data_a.csv", data_a, 5, 0, 0, 0},
        {"data_b.csv", data_b, 6, 0, 0, 0},
        {"data_c.csv", data_c, 3, 0, 0, 0}
    };

    pthread_t tasks[3];
    for (int i = 0; i < 3; i++) pthread_create(&tasks[i], NULL, process_file, &results[i]);
    for (int i = 0; i < 3; i++) pthread_join(tasks[i], NULL);

    for (int i = 0; i < 3; i++) {
        printf("%s: sum=%d max=%d count=%d\n",
               results[i].name, results[i].sum, results[i].max, results[i].count);
    }
    printf("all files processed\n");
    return 0;
}

File: source/go/async_file_processing.go
package main

import (
	"fmt"
	"sync"
)

type FileResult struct {
	name  string
	sum   int
	max   int
	count int
}

func processFile(name string, data []int, resultCh chan<- FileResult, wg *sync.WaitGroup) {
	defer wg.Done()
	sum := 0
	max := data[0]
	for _, v := range data {
		sum += v
		if v > max {
			max = v
		}
	}
	resultCh <- FileResult{name: name, sum: sum, max: max, count: len(data)}
}

func main() {
	type fileData struct {
		name string
		data []int
	}
	files := []fileData{
		{"data_a.csv", []int{10, 20, 30, 40, 50}},
		{"data_b.csv", []int{5, 15, 25, 35, 45, 55}},
		{"data_c.csv", []int{100, 200, 300}},
	}

	resultCh := make(chan FileResult, len(files))
	var wg sync.WaitGroup

	for _, f := range files {
		wg.Add(1)
		go processFile(f.name, f.data, resultCh, &wg)
	}

	wg.Wait()
	close(resultCh)

	results := make([]FileResult, 0, len(files))
	for r := range resultCh {
		results = append(results, r)
	}

	// Sort results by name for deterministic output
	for i := 0; i < len(results); i++ {
		for j := i + 1; j < len(results); j++ {
			if results[i].name > results[j].name {
				results[i], results[j] = results[j], results[i]
			}
		}
	}

	for _, r := range results {
		fmt.Printf("%s: sum=%d max=%d count=%d\n", r.name, r.sum, r.max, r.count)
	}
	fmt.Println("all files processed")
}
